server/src/test/java/org/opensearch/cluster/routing/allocation/NodeLoadAwareAllocationTests.java

```diff
@@ -0,0 +1,1220 @@
+        newState = strategy.reroute(newState, "reroute");
+        while (newState.getRoutingNodes().shardsWithState(INITIALIZING).isEmpty() == false) {
+            newState = startInitializingShardsAndReroute(strategy, newState);
+        }
+        assertThat(newState.getRoutingNodes().shardsWithState(STARTED).size(), equalTo(32));
+
+        //add back node2 when skewness is still breached
+        newState = ClusterState.builder(newState).nodes(DiscoveryNodes.builder(newState.nodes())
+            .add(newNode("node2", singletonMap("zone", "zone_1"))))
+            .build();
+        newState = strategy.reroute(newState, "reroute");
+        while (newState.getRoutingNodes().shardsWithState(INITIALIZING).isEmpty() == false) {
+            newState = startInitializingShardsAndReroute(strategy, newState);
+        }
+        assertThat(newState.getRoutingNodes().shardsWithState(STARTED).size(), equalTo(36));
+
+        //add back node3
+        newState = ClusterState.builder(newState).nodes(DiscoveryNodes.builder(newState.nodes())
+            .add(newNode("node3", singletonMap("zone", "zone_1"))))
+            .build();
+        newState = strategy.reroute(newState, "reroute");
+
+        while (newState.getRoutingNodes().shardsWithState(INITIALIZING).isEmpty() == false) {
+            newState = startInitializingShardsAndReroute(strategy, newState);
+        }
+        assertThat(newState.getRoutingNodes().shardsWithState(STARTED).size(), equalTo(40));
+    }
+
+    public void testExistingPrimariesAllocationOnOverload() {
+        GatewayAllocator gatewayAllocator = new TestGatewayAllocator();
+        AllocationService strategy = createAllocationService(Settings.builder()
+            .put(ThrottlingAllocationDecider.CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey(), 20)
+            .put(ThrottlingAllocationDecider.CLUSTER_ROUTING_ALLOCATION_NODE_INITIAL_PRIMARIES_RECOVERIES_SETTING.getKey(), 20)
+            .put(ThrottlingAllocationDecider.CLUSTER_ROUTING_ALLOCATION_NODE_INITIAL_REPLICAS_RECOVERIES_SETTING.getKey(), 20)
+            .put(ClusterRebalanceAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ALLOW_REBALANCE_SETTING.getKey(), "always")
+            .put(NodeLoadAwareAllocationDecider.CLUSTER_ROUTING_ALLOCATION_LOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING.getKey(), 5)
+            .put(NodeLoadAwareAllocationDecider.CLUSTER_ROUTING_ALLOCATION_LOAD_AWARENESS_SKEW_FACTOR_SETTING.getKey(), 50)
+            .put(NodeLoadAwareAllocationDecider.CLUSTER_ROUTING_ALLOCATION_LOAD_AWARENESS_ALLOW_UNASSIGNED_PRIMARIES_SETTING.getKey(),
+                false)
+            .build(), gatewayAllocator);
+
+        logger.info("Building initial routing table for 'testExistingPrimariesAllocationOnOverload'");
+
+        Metadata metadata = Metadata.builder()
+            .put(IndexMetadata.builder("test").settings(settings(Version.CURRENT)).numberOfShards(20).numberOfReplicas(0))
+            .build();
```

Comment:
Thanks for adding this test. This is useful. Do you think having replicas as well, would have demonstrated that only primaries are being force allocated, while replicas remain unassigned when the nodes join back.

Reply:
Thats a test in for the `canForceAllocatePrimary` since there is nothing spl that this code does for 
>while replicas remain unassigned when the nodes join back

Reply:
To assert on the exclusiveness of the behaviour which applies to existing primary but not to existing replica. If that is covered elsewhere it is fine.